### ISSUE TITLE
Add .gfa output to racon in addition to .fasta

### ISSUE THREAD

**Issue by jelber2** [isovic/racon#49]
Hi, I was wondering if it were possible for racon to output a gfa file in addition to fasta?

**Comment by rvaser** [comment]
Hello,
what would be the use case of the outputed gfa format (i.e. what information are you seeking from racon)?

Best regards,
Robert

**Comment by jelber2** [comment]
Well, I would like to input a gfa file from racon to the Hi-C scaffolding program SALSA (https://github.com/machinegun/SALSA). Granted I don't know if a gfa representation of the assembly would improve the output from SALSA or not.

**Comment by rvaser** [comment]
I have tagged this as enhancement and will adress it soon.

Best regards,
Robert

**Comment by sjackman** [comment]
I'd find this feature useful as well. I'm polishing a Miniasm assembly using Racon. It'd be useful to preserve the graph after polishing with Racon. Consider supporting both GFA 1 and GFA 2.

**Comment by rvaser** [comment]
How should I preserve the GFA file? Sequences change and alignments might be invalidate.

**Comment by sjackman** [comment]
The GFA 1 output by Miniasm includes estimates of the amount of overlap, but doesn't include an actual alignment. So I think you could get away with not modifying the edges at all. The edges output by Miniasm look like this:
```
L	utg000001l	+	utg001226l	+	19386M	SD:i:5467
```

After the sequences are corrected by Racon, you could realign the two sequences incident to each edge, and it's possible that some of the ambiguities in the graph could be resolved post-Racon.

**Comment by rvaser** [comment]
It is a bit tedious to add the format into Racon as we only need the S rows. Wouldn't a simple post-processing script be an easier solution? A script that updates the GFA file with polished sequences and maybe realigns edges?

**Comment by sjackman** [comment]
A post-processing script may be easiest. That script would take in the GFA file produced by Miniasm, the FASTA file produced by Racon, and produce an updated GFA file. Is that script something that you're interested in creating? Or perhaps a task for Gfakluge or GfaPy.

**Comment by rvaser** [comment]
Well I might add such a script but I am not sure when I will get the time for it :/

**Comment by sjackman** [comment]
No worries. I'll let you know if I get around to it myself.

**Comment by rvaser** [comment]
Great, thanks!

**Comment by mictadlo** [comment]
Any updates?

**Comment by sjackman** [comment]
Not from me

**Comment by rvaser** [comment]
Neither from me :/

**Comment by SamStudio8** [comment]
I don't suppose anyone had a chance to look at this?

**Comment by rvaser** [comment]
Unfortunately not :/ I'll try and deal with it later this year.

**Comment by sjackman** [comment]
I used this AWK script to take the sequence from `polished.fasta` the graph from `draft.gfa` and produce a `polished.gfa` file.
```sh
seqtk seq polished.fasta | gawk -vOFS='\t' 'ARGIND == 1 { id = substr($1, 2); getline; x[id] = $1; next } $1 == "S" && x[$2] { $3 = x[$2] } 1' - draft.gfa >polished.gfa
```
See also https://github.com/edawson/gfakluge and https://github.com/ggonnella/gfapy/ for manipulating GFA files.
I'd still love to see an option in Racon for this purpose.

**Comment by MChiaraC** [comment]
so you basically taking the unpolished assembly graph and the new polished sequences and creating a polished graph? Am I correct?

**Comment by jelber2** [comment]
That what I understand @sjackman's code is doing.

**Comment by sjackman** [comment]
Yes. I'm working with an assembly graph whose edges are blunt (no overlap, `0M`) from Flye or Unicycler. This simple script does not recompute the edge alignment for other assemblers.

**Comment by MChiaraC** [comment]
mmmh I see, than I cannot use it ...

**Comment by sjackman** [comment]
You could replace all the CIGAR strings with `*` (meaning unknown).

**Comment by ardy20** [comment]
Hello Robert
Any progress or update to create .gfa output by Racon?

**Comment by jelber2** [comment]
See https://github.com/rrwick/Minipolish

**Comment by rvaser** [comment]
@ardy20, unfortunately no. Minipolish seems as a decent solution for this issue :)
